### PR TITLE
#277 - reply to messages

### LINKED ISSUE #277: [Admin Panel] Reply to messages via administration panel
Administrator should be able to reply to the message (send an email) from contact form via the administration panel

### PR DESCRIPTION
**Pull request by AleksandraKozubal**
Created modal in admin panel with possibility to respond on an email and simple email layout.
It should close #277.

image
image

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -32,12 +32,12 @@ REDIS_PASSWORD=null
...
-MAIL_HOST=mailpit
-MAIL_PORT=1025
+MAIL_HOST=smtp.gmail.com
+MAIL_PORT=587
 MAIL_USERNAME=null
 MAIL_PASSWORD=null
-MAIL_ENCRYPTION=null
-MAIL_FROM_ADDRESS="hello@example.com"
+MAIL_ENCRYPTION=tls
+MAIL_FROM_ADDRESS="no-reply@blumilk.pl"
```
Comment (kamilpiech97) [inline .env.example:40]: Why `gmail`? Locally, we are using mailpit.

File: app/Filament/Resources/ContactFormResource.php [hunk 2]
```diff
@@ -50,6 +51,12 @@ public static function form(Form $form): Form
                             ->rows(4)
                             ->label("Wiadomość")
                             ->disabled(),
+                        Forms\Components\RichEditor::make("response")
+                            ->label("Odpowiedź")
+                            ->disableToolbarButtons(["attachFiles"])
+                            ->maxLength(65000)
+                            ->disabled()
+                            ->visible(fn($record) => $record->status === ContactFormStatus::Responded),
```
Comment (kamilpiech97) [inline app/Filament/Resources/ContactFormResource.php:59]: Typing

File: app/Filament/Resources/ContactFormResource.php [hunk 3]
```diff
@@ -105,7 +112,8 @@ public static function table(Table $table): Table
             ->actions([
                 Tables\Actions\ViewAction::make(),
                 Tables\Actions\EditAction::make(),
-                Tables\Actions\DeleteAction::make(),
+                Tables\Actions\DeleteAction::make()->visible(fn($record) => $record->status !== ContactFormStatus::Responded),
```
Comment (kamilpiech97) [inline app/Filament/Resources/ContactFormResource.php:115]: Typing

**Review by kamilpiech97 — CHANGES_REQUESTED** [review]

File: .env.example (hunk 1 above)
Comment (AleksandraKozubal) [reply .env.example]: ok, changed back

**Review by kamilpiech97 — DISMISSED** [review]

File: resources/views/email.blade.php [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+                    <table role="presentation" style="margin: 0; text-align: left; width: 100%;">
+                        <tr>
+                            <td style="font-size: 16px; text-align: left;" class="content-block">
+                                <div style="display: flex; align-items: center;">
+                                    <img src="https://www.blumilk.pl/logo.png" class="logo" style="height: 80px; margin: auto 8px auto 16px;" alt="Blumilk Logo">
+                                    <div>
+                                        <span>Blumilk sp. z o.o.</span><br>
+                                        <span style="color: #000000; text-decoration: none;">office@blumilk.pl</span><br>
+                                        <span>+48 780 142 367</span><br>
+                                        <span>ul. Najświętszej Marii Panny 5f/5 59-220 Legnica</span><br>
```
Comment (EwelinaSkrzypacz) [inline resources/views/email.blade.php:41]:
```suggestion
                                        <span>ul. Najświętszej Marii Panny 5F/5 59-220 Legnica</span><br>
```

I think that on contact page is the same little typo

File: resources/views/email.blade.php [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+                    </table>
+                </div>
+            </div>
+        </td>
+        <td>&nbsp;</td>
+    </tr>
+</table>
+</body>
+</html>
+
```
Comment (EwelinaSkrzypacz) [inline resources/views/email.blade.php:62]:
```suggestion
```

**Review by EwelinaSkrzypacz — COMMENTED** [review]
When message status is "Odpowiedziano" on index page the button "Usuń" is hidden

image


But on edit page there is button "Usuń" and after clicking on it, the message is deleted. 

image

**Comment by AleksandraKozubal** [comment]
> When message status is "Odpowiedziano" on index page the button "Usuń" is hidden
> 
> image
> 
> But on edit page there is button "Usuń" and after clicking on it, the message is deleted.
> 
> image

fixed

**Review by EwelinaSkrzypacz — COMMENTED** [review]
Our footer in mails looks like this:

image

I mean that we have logo with company name and also link to company's page. 

I don't know how much should we focus on e-mail design

**Comment by AleksandraKozubal** [comment]
> Our footer in mails looks like this:
> 
> image
> 
> I mean that we have logo with company name and also link to company's page.
> 
> I don't know how much should we focus on e-mail design

What do you think about leaving it like it is, writing note about email design in #285? If we want, we can ask designer to design it

**Review by EwelinaSkrzypacz — DISMISSED** [review]

File: app/Filament/Resources/ContactFormResource/Actions/RespondToContactFormMessageAction.php [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+class RespondToContactFormMessageAction extends Action
+{
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        $this->label("Odpowiedz")
+            ->icon("heroicon-m-envelope")
+            ->color("success")
+            ->visible(fn($record) => $record->status !== ContactFormStatus::Responded)
```
Comment (kamilpiech97) [inline app/Filament/Resources/ContactFormResource/Actions/RespondToContactFormMessageAction.php:23]: typing

File: app/Filament/Resources/ContactFormResource/Actions/RespondToContactFormMessageAction.php [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+                    ->disableToolbarButtons(["attachFiles", "codeBlock", "blockquote"])
+                    ->maxLength(65000)
+                    ->required(),
+            ])
+            ->action(function ($record, $data): void {
+                $details = [
+                    "subject" => $data["responseTopic"],
+                    "response" => $data["response"],
+                    "messageDescription" => $record->message,
+                ];
```
Comment (kamilpiech97) [inline app/Filament/Resources/ContactFormResource/Actions/RespondToContactFormMessageAction.php:53]: Maybe we should use DTO here?

Comment (kamilpiech97) [reply app/Filament/Resources/ContactFormResource/Actions/RespondToContactFormMessageAction.php:53]: `$data` looks ugly

File: app/Mail/ContactFormResponded.php [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+{
+    use Queueable;
+    use SerializesModels;
+
+    public array $details;
+
+    public function __construct($details)
+    {
+        $this->details = $details;
+    }
```
Comment (kamilpiech97) [inline app/Mail/ContactFormResponded.php:21]: https://stitcher.io/blog/constructor-promotion-in-php-8

**Review by kamilpiech97 — CHANGES_REQUESTED** [review]

File: app/Filament/Resources/ContactFormResource/Actions/RespondToContactFormMessageAction.php (hunk 7 above)
Comment (AleksandraKozubal) [reply app/Filament/Resources/ContactFormResource/Actions/RespondToContactFormMessageAction.php]: used that

File: app/Filament/Resources/ContactFormResource/Actions/RespondToContactFormMessageAction.php [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+                    ->maxLength(65000)
+                    ->required(),
+            ])
+            ->action(function (ContactForm $record, array $data): void {
+                $details = ContactFormResponse::fromArray($record, $data);
+
+                Mail::to($record->email)->send(new ContactFormResponded($details));
+
+                $record->update([
+                    "response" => $data["response"],
```
Comment (kamilpiech97) [inline app/Filament/Resources/ContactFormResource/Actions/RespondToContactFormMessageAction.php:56]:
```suggestion
                    "response" => $details->response,
```
why not like that?

**Review by kamilpiech97 — APPROVED** [review]

**Review by EwelinaSkrzypacz — APPROVED** [review]

**Pull request merged**